lib/cadet/assessments/assessments.ex

```diff
@@ -371,6 +372,73 @@ defmodule Cadet.Assessments do
     end
   end
 
+  def unsubmit_submission(submission_id, avenger = %User{role: role})
+      when is_ecto_id(submission_id) do
+    if role in @unsubmit_assessment_role do
+      submission =
+        Submission
+        |> join(:inner, [s], a in assoc(s, :assessment))
+        |> preload([_, a], assessment: a)
+        |> Repo.get(submission_id)
+
+      with {:submission_found?, true} <- {:submission_found?, is_map(submission)},
+           {:is_open?, true} <- is_open?(submission.assessment),
+           {:status, :submitted} <- {:status, submission.status},
+           {:avenger_of?, true} <-
+             {:avenger_of?, Cadet.Accounts.Query.is_avenger_of(avenger, submission.student_id)} do
+        Ecto.Multi.new()
+        |> Multi.run(
+          :rollback_submission,
+          fn _ ->
+            submission
+            |> Submission.changeset(%{status: :attempted, xp_bonus: 0})
+            |> Repo.update()
+          end
+        )
+        |> Multi.run(:rollback_answers, fn _ ->
+          Answer
+          |> where(submission_id: ^submission.id)
+          |> Repo.all()
+          |> Enum.each(fn answer ->
```

Comment:
Instead I think you should instead do `Enum.reduce` on the result of `Repo.update/1` and then make sure the function passed to `Multi.run/3` returns the appropriate `{:ok, nil}` or `{:error, ...}` because right now your multi will not terminate and rollback on update error.